lib/ansible/modules/storage/glusterfs/gluster_peer.py

```diff
@@ -0,0 +1,181 @@
+         force: true
+'''
+
+RETURN = '''
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ast import literal_eval
+
+
+class Peer(object):
+    def __init__(self, module):
+        self.module = module
+        self.state = self._validated_params('state')
+
+    def _validated_params(self, opt):
+        if self.module.params[opt] is None:
+            msg = "Please provide %s option in the playbook!" % opt
+            self.module.fail_json(rc=4, msg=msg)
+        return self.module.params[opt]
+
+    def get_nodes(self):
+        nodes = literal_eval(self._validated_params('nodes'))
+        if nodes[0] is None:
+            # Found a list with None as its first element
+            self.module.fail_json(rc=1, msg="Nodes not found, provide at least "
+                                  "one node.")
+        return nodes
+
+    def gluster_peer_ops(self):
+        nodes = self.get_nodes()
+        force = 'force' if self.module.params.get('force') else ''
+        if self.state == 'present':
+            nodes = self.get_to_be_probed_hosts(nodes)
+            action = 'probe'
+        elif self.state == 'absent':
+            action = 'detach'
+        cmd = []
+        for node in nodes:
+            cmd.append(' peer ' + action + ' ' + node + ' ' + force)
+        return cmd
+
+    def get_to_be_probed_hosts(self, hosts):
+        rc, output, err = self.module.run_command("gluster pool list")
+        peers_in_cluster = [line.split('\t')[1].strip() for
+                            line in filter(None, output.split('\n')[1:])]
+        try:
+            peers_in_cluster.remove('localhost')
+        except ValueError:
+            # It is ok not to have localhost in list
+            pass
+        hosts_to_be_probed = [host for host in hosts if host not in
+                              peers_in_cluster]
+        return hosts_to_be_probed
+
+    def call_peer_commands(self, cmds):
+        errors = []
+        for cmd in cmds:
+            rc, out, err = self._run_command('gluster', cmd)
+            if rc:
+                errors.append(err)
+        return errors
+
+    def get_output(self, errors):
+        if not errors:
+            self.module.exit_json(rc=0, changed=True)
+        else:
+            self.module.fail_json(rc=1, msg='\n'.join(errors))
+
+    def _run_command(self, op, opts):
+        cmd = self.module.get_bin_path(op, True) + opts + ' --mode=script'
+        return self.module.run_command(cmd)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            force=dict(type='bool', required=False),
+            nodes=dict(),
+            state=dict(required=True, choices=["absent", "present"]),
+        ),
+        supports_check_mode=False,
+    )
+    pops = Peer(module)
+    # Verify if GlusterFS 3.2 or over is installed
+    if not valid_gluster_version(module):
+        module.fail_json(msg="GlusterFS version > 3.2 is required")
+    cmds = pops.gluster_peer_ops()
```

Comment:
`gluster_peer_ops` is unconditionally called and this method always calls `get_nodes`. `get_nodes` fails when `nodes` parameter isn't set: this parameter must be mandatory (and non empty).
For that, you could use a custom method, meaning something like that:
```
class AnsibleModuleCheckListNotEmpty(AnsibleModule):
    def _check_type_list_not_empty(self, value):
        value = self._check_type_list(value) # default checks for a list
        if not value:
            raise ValueError("list must not be empty")
        return value
[...]
module = AnsibleModuleCheckListNotEmpty(
        argument_spec=dict(
            force=dict(type='bool', required=False),
            nodes=dict(type=self._check_type_list_not_empty, required=True),
[...]
```
Once implemented you could remove the `get_nodes` method.

Reply:
@pilou- I somehow couldn't get this working. I was hitting implementation error, web or the other ansible module sources didn't give me much clue.

However I have tried another approach, that is again if-else statements. I have made `nodes' a required parameter and anyway end up calling gluster_peer_ops. I am committing for review.

Can you please point me to some documentation or source files?

Reply:
callable type is used |there](https://github.com/ansible/ansible/blob/2ffbd5554f50bb8e087d3d2b94d0ac33a0790778/lib/ansible/module_utils/basic.py#L2038)
It is tested [there](https://github.com/ansible/ansible/blob/2ffbd5554f50bb8e087d3d2b94d0ac33a0790778/test/units/module_utils/basic/test_argument_spec.py#L160)
It doesn't seem documented. It is mentioned in 2.4 [changelog](https://github.com/ansible/ansible/blob/stable-2.4/CHANGELOG.md):
> Python module argument_spec now supports custom validation logic by accepting a callable as the type argument

Another way to do: specify `type=list` for `nodes` parameter ([sample](https://github.com/ansible/ansible/blob/2ffbd5554f50bb8e087d3d2b94d0ac33a0790778/lib/ansible/modules/system/user.py#L2128)), remove the call to `literal_eval` and keep the `if not nodes` check.

Reply:
I couldn't get the custom method working. I end up at `if not callable', https://github.com/ansible/ansible/blob/2ffbd5554f50bb8e087d3d2b94d0ac33a0790778/lib/ansible/module_utils/basic.py#L2049 maybe I'm missing something obvious. 

Used type=list instead and got rid of literal_eval, wonder why I didn't use it earlier.

Thanks for the review.